File: 03-Artificial-Neural-Networks/csharp-neural-network/MnistSamples.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharp_Neural_Network
{
    class MnistSamples
    {
        public Sample[] Training { get; private set; }
        public Sample[] Testing { get; private set; }

        public MnistSamples()
        {
            var sourceDir = new DirectoryInfo(AppDomain.CurrentDomain.BaseDirectory).Parent.Parent;
            var dataDir = Path.Combine(sourceDir.FullName, "MnistData");
            var dataFile = Path.Combine(dataDir, "Training.csv");

            var allSamples = File.ReadAllLines(dataFile)
                .Skip(1)
                .Select(line => new Sample(
                    Sample.GetValues(line.Substring(2)),
                    GetTargetArray(line.Substring(0, 1))))
                .ToArray();
            var testCount = allSamples.Length / 10;
            Training = new Sample[allSamples.Length - testCount];
            Array.Copy(allSamples, Training, Training.Length);
            Testing = new Sample[testCount];
            if (testCount > 0)
                Array.Copy(allSamples, Training.Length, Testing, 0, Testing.Length);
        }

        static double[] GetTargetArray(string targetText)
        {
            var targetValue = int.Parse(targetText);
            var targetArray = new double[10];
            for (int i = 0; i < 10; i++)
                targetArray[i] = i == targetValue ? 1 : 0;
            return targetArray;
        }

    }
}

File: 03-Artificial-Neural-Networks/fsharp-neural-network2/MnistSamples.fs
module MnistSamples

open System
open System.IO
open Sample

let mnistSamples() = 
    let sourceDir = (new DirectoryInfo(AppDomain.CurrentDomain.BaseDirectory)).Parent.Parent
    let dataDir = Path.Combine(sourceDir.FullName, "MnistData");
    let dataFile = Path.Combine(dataDir, "Training.csv");

    let getTargetArray targetText = 
        let target = int targetText
        [|0..9|] |> Array.map (fun i -> if i = target then 1. else 0.)

    let allSamples =
        File.ReadAllLines(dataFile)
        |> Seq.skip 1
        |> Seq.map (fun line -> 
           { Input = (getFloatsFromText (line.Substring 2)) 
             Target = (getTargetArray (line.Substring(0, 1))) })
        |> Seq.toArray
    let testCount = allSamples.Length / 10
    let firstTest = allSamples.Length - testCount
    allSamples.[..(firstTest - 1)], allSamples.[firstTest..]
    
            

